### PR TITLE
[#2552] feat(catalog-jdbc-doris): add Doris container implement for integration-test

### PR DESCRIPTION
**Pull request by zhoukangcn**
### What changes were proposed in this pull request?

add Doris container implement for integration-test

### Why are the changes needed?

Fix: #2552 

### Does this PR introduce _any_ user-facing change?

N/A

### How was this patch tested?

UT

### CODE REVIEW

File: integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            break;
+          }
+        }
+
+        if (!isDorisContainerReady) {
+          LOG.info("Doris container is not ready yet!");
+          Thread.sleep(5000);
+        } else {
+          break;
+        }
```
Comment (yuqi1129) [inline integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java:92]:
```suggestion
if (isDorisContainerReady) {
    return true;
}

LOG.info("Doris container is not ready yet!");
Thread.sleep(5000);

```

File: integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    super.setupContainer();
+    withLogConsumer(new PrintingContainerLog(format("%-14s| ", "DorisContainer")));
+    withStartupTimeout(Duration.ofMinutes(10));
+  }
+
+  @Override
+  public void start() {
+    super.start();
+    Preconditions.check("Doris container startup failed!", checkContainerStatus(5));
+    changePassword();
```
Comment (yuqi1129) [inline integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java:59]: If we fail to change the password, I believe we should throw an exception and stop the test.

File: integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2023 Datastrato Pvt Ltd.
```
Comment (yuqi1129) [inline integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java:2]: 2024

Comment (zhoukangcn) [reply integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java]: done

File: integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java (hunk 2 above)
Comment (zhoukangcn) [reply integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java]: done. Use `Preconditions.check` instead

File: integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java (hunk 1 above)
Comment (zhoukangcn) [reply integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java]: done

File: catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/TestDorisDatabaseOperations.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+@Tag("gravitino-docker-it")
+public class TestDorisDatabaseOperations extends TestDorisAbstractIT {
+
+  @Test
+  public void testBaseOperationDatabase() {
+    String databaseName = RandomNameUtils.genRandomName("ct_db");
+    Map<String, String> properties = new HashMap<>();
+    String databaseComment = "test_comment";
+
+    Assertions.assertThrows(
```
Comment (mchades) [inline catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/TestDorisDatabaseOperations.java:23]: Could you please also assert the exception message?

File: catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/TestDorisDatabaseOperations.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        UnsupportedOperationException.class,
+        () -> {
+          DATABASE_OPERATIONS.create(databaseName, databaseComment, properties);
```
Comment (mchades) [inline catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/TestDorisDatabaseOperations.java:26]: It seems a UT not an e2e test, but you put this in the package `integration.test`

File: integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+      String dorisMySQLUrl = format("jdbc:mysql://%s:%d/", getContainerIpAddress(), FE_MYSQL_PORT);
+      Connection connection = DriverManager.getConnection(dorisMySQLUrl, USER_NAME, "");
+
+      String query = String.format("SET PASSWORD FOR '%s' = PASSWORD('%s');", USER_NAME, PASSWORD);
+      Statement statement = connection.createStatement();
+      statement.execute(query);
+    } catch (Exception e) {
+      LOG.info(e.getMessage(), e);
+      return false;
+    }
```
Comment (mchades) [inline integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java:112]: The `connection` need to be closed?

File: integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+      Map<String, String> envVars,
+      Optional<Network> network) {
+    super(image, hostName, ports, extraHosts, filesToMount, envVars, network);
+  }
+
+  @Override
+  protected void setupContainer() {
+    super.setupContainer();
+    withLogConsumer(new PrintingContainerLog(format("%-14s| ", "DorisContainer")));
+    withStartupTimeout(Duration.ofMinutes(10));
```
Comment (mchades) [inline integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java:52]: why set the timeout as 10 minutes? Seems a bit long

File: catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/TestDorisDatabaseOperations.java (hunk 5 above)
Comment (yuqi1129) [reply catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/TestDorisDatabaseOperations.java:26]: It's normal, currently, we don't differentiate between IT and UT very much.  The package `integration.test` holds tests that rely on the docker environment.

File: integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java (hunk 7 above)
Comment (zhoukangcn) [reply integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java:52]: seems 10 minutes is not enough. Doris Docker image size is too big, docker pull for github CI will cost too many time.

Maybe I'd better use old Doris version to reduce docker image size

File: catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/TestDorisDatabaseOperations.java (hunk 5 above)
Comment (yuqi1129) [reply catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/TestDorisDatabaseOperations.java:26]: Yeah, `TestDorisDatabaseOperations` relies on docker environment but it's a unit test actually.

Comment (mchades) [reply catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/TestDorisDatabaseOperations.java:26]: Therefore, I do not recommend putting this class in the package `integration.test`

File: catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/TestDorisDatabaseOperations.java (hunk 4 above)
Comment (zhoukangcn) [reply catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/TestDorisDatabaseOperations.java:23]: [DorisDatabaseOperations](https://github.com/datastrato/gravitino/blob/main/catalogs/catalog-jdbc-doris/src/main/java/com/datastrato/gravitino/catalog/doris/operation/DorisDatabaseOperations.java#L24)

`throw new UnsupportedOperationException();`

so, maybe it's no need to assert the exception message

File: integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java (hunk 6 above)
Comment (zhoukangcn) [reply integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java:133]: done

File: catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/TestDorisDatabaseOperations.java (hunk 5 above)
Comment (zhoukangcn) [reply catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/TestDorisDatabaseOperations.java:26]: In other projects (like Apache kyuubi), they also use Docker container in unit test. so I think it's ok here.

https://github.com/apache/kyuubi/blob/master/externals/kyuubi-jdbc-engine/src/test/scala/org/apache/kyuubi/engine/jdbc/doris/WithDorisContainer.scala

Comment (yuqi1129) [reply catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/TestDorisDatabaseOperations.java:26]: > Therefore, I do not recommend putting this class in the package `integration.test`

At this point, I discussed it with @jerryshao in #2279 , so it is acceptable for the package "integration.test" to hold tests that are dependent upon the docker environment.

File: integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java (hunk 7 above)
Comment (yuqi1129) [reply integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java:53]: Based on this scenario, I believe we may need to cache the image in CI. see https://github.com/datastrato/gravitino/issues/998 or https://github.com/datastrato/gravitino/issues/1142

File: catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/TestDorisDatabaseOperations.java (hunk 5 above)
Comment (yuqi1129) [reply catalogs/catalog-jdbc-doris/src/test/java/com/datastrato/gravitino/catalog/doris/integration/test/TestDorisDatabaseOperations.java:26]: We will use another issue https://github.com/datastrato/gravitino/issues/2599 to track it, @zhoukangcn please go ahead.

File: integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java (hunk 7 above)
Comment (yuqi1129) [reply integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java:53]: According to the log, the root cause is that the service can't be launched within 10 minutes, please check it again
[catalog-jdbc-doris-integration-test.log](https://github.com/datastrato/gravitino/files/14681505/catalog-jdbc-doris-integration-test.log)

File: integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+      }
+    }
+
+    if (connection != null) {
+      try {
+        connection.close();
+      } catch (SQLException e) {
+        LOG.warn(e.getMessage(), e);
+      }
+    }
```
Comment (mchades) [inline integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java:105]: You should use try-with-resource to ensure that the connection is properly closed after use. For example:
```java
protected boolean checkContainerStatus(int retryLimit) {
    String dorisMySQLUrl = format("jdbc:mysql://%s:%d/", getContainerIpAddress(), FE_MYSQL_PORT);
    LOG.info("Doris url is " + dorisMySQLUrl);

    for (int nRetry = 0; nRetry < retryLimit; nRetry++) {
        try (Connection connection = DriverManager.getConnection(dorisMySQLUrl, USER_NAME, "")) {
             Statement statement = connection.createStatement()) {

            String query = "SHOW PROC '/backends';";
            ResultSet resultSet = statement.executeQuery(query);
            while (resultSet.next()) {
                String alive = resultSet.getString("Alive");
                if ("true".equalsIgnoreCase(alive)) {
                    LOG.info("Doris container startup success!");
                    return true;
                }
            }

            LOG.info("Doris container is not ready yet!");
            Thread.sleep(5000);
        } catch (SQLException | InterruptedException e) {
            LOG.error(e.getMessage(), e);
            // ...
        }
    }

    return false;
}
```

File: integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java (hunk 7 above)
Comment (zhoukangcn) [reply integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java:53]: @yuqi1129 thanks, I will fix it later.

File: dev/ci/util_free_space.sh [hunk 9]
```diff
@@ -0,0 +1,57 @@
+#!/bin/bash
+#
```
Comment (qqqttt123) [inline dev/ci/util_free_space.sh:2]: We can't use this license header. This file is from Apache arrow. We should add license information in License file.

Comment (zhoukangcn) [reply dev/ci/util_free_space.sh:2]: Thank you for the reminder. This file is here for testing purposes. I will move it to another PR later and update LICENSE

Comment (zhoukangcn) [reply dev/ci/util_free_space.sh]: This file is deleted in this PR

**Comment by zhoukangcn** [comment]
@yuqi1129 @mchades @qqqttt123   I had fix integration test, and update this PR. 
Please help to review this PR again when you have time.

File: .github/workflows/backend-integration-test.yml [hunk 10]
```diff
@@ -55,7 +55,7 @@ jobs:
     needs: changes
     if: needs.changes.outputs.source_changes == 'true'
     runs-on: ubuntu-latest
-    timeout-minutes: 30
+    timeout-minutes: 60
```
[the remark this answers is not in the document]
Comment (zhoukangcn) [reply .github/workflows/backend-integration-test.yml:58]: backend integration test need 42 minutes now. So, I increase the timeout.

File: integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+    String dorisJdbcUrl = format("jdbc:mysql://%s:%d/", getContainerIpAddress(), FE_MYSQL_PORT);
+    LOG.info("Doris url is " + dorisJdbcUrl);
+
+    while (nRetry++ < retryLimit) {
+      try (Connection connection = DriverManager.getConnection(dorisJdbcUrl, USER_NAME, "")) {
+
+        // execute `SHOW PROC '/backends';` to check if backends is ready
+        String query = "SHOW PROC '/backends';";
+        Statement statement = connection.createStatement();
```
Comment (yuqi1129) [inline integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java:75]: Maybe you'd better close `Statement` as well. 

The same for line 107.

File: integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+        ResultSet resultSet = statement.executeQuery(query);
+        while (resultSet.next()) {
+          String alive = resultSet.getString("Alive");
+          if (alive.equalsIgnoreCase("true")) {
+            LOG.info("Doris container startup success!");
+            isDorisContainerReady = true;
```
Comment (yuqi1129) [inline integration-test-common/src/test/java/com/datastrato/gravitino/integration/test/container/DorisContainer.java:81]: I would suggest directly returning `true`, the same goes for line 87.

**Review by yuqi1129 — APPROVED** [review]
LGTM

**Comment by zhoukangcn** [comment]
@mchades do you have any further advice?

**Review by mchades — APPROVED** [review]
LGTM

**Comment by yuqi1129** [comment]
Thanks for your hard work.

**Pull request merged**